### PR TITLE
feat(tx_index): fill transaction index gap on enabling automatic backfill

### PR DESCRIPTION
**Pull request by akaladarshi**
## Related Issues
Fixes part of: #11007 

## Proposed Changes
- Adds `EnableAutomaticBackFill` in index config in `fullNode`
- Adds `MaxAutomaticBackFillTxIndexHeight` in index config in `fullnode`
- Tries to fill the gap between the last populated from `current tipset tx hashes` till:
  - we reach `MaxAutomaticBackFillTxIndexHeight` or
  - we reach a conflict in the `eth tx hash index` database
- Adds `PopulateAfterSnapshot` for eth tx hash index to populate data for while loading from snapshot

### CODE REVIEW

**Comment by akaladarshi** [comment]
@aarshkshah1992 As per our discussion about waiting till the observer gets the latest head and process that,
I have a few ideas, let's discuss them once we verify this is correct.

File: node/modules/ethmodule.go [hunk 1]
```diff
@@ -78,7 +84,7 @@ func EthModuleAPI(cfg config.FevmConfig) func(helpers.MetricsCtx, repo.LockedRep
 				}
 
 				// Tipset listener
-				_ = ev.Observe(&ethTxHashManager)
+				ev.Observe(&ethTxHashManager)
```
Comment (aarshkshah1992) [inline node/modules/ethmodule.go:88]: We don't need this change.

File: node/config/types.go [hunk 2]
```diff
@@ -621,6 +621,9 @@ type IndexConfig struct {
 	// EXPERIMENTAL FEATURE. USE WITH CAUTION
 	// EnableMsgIndex enables indexing of messages on chain.
 	EnableMsgIndex bool
+
+	// EnableAutomaticBackFill enables automatic index back-filling
+	EnableAutomaticBackFill bool
```
Comment (aarshkshah1992) [inline node/config/types.go:626]: For now, we need to rename this to `EnableAutomaticBackFillTxIndex` as we're only doing it for the TX Index.

File: node/impl/full/txhashmanager.go [hunk 3]
```diff
@@ -23,6 +23,49 @@ func (m *EthTxHashManager) Revert(ctx context.Context, from, to *types.TipSet) e
...
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Finished filling transaction index gap. Total processed block: %d", totalProcessedBlock)
+
+	return nil
+}
+
+// PopulateExistingMappings walk back from the current head to the minimum height and populate the eth transaction hash lookup database
```
Comment (aarshkshah1992) [inline node/impl/full/txhashmanager.go:68]: nit: check grammar

File: node/impl/full/txhashmanager.go [hunk 4]
```diff
@@ -23,6 +23,49 @@ func (m *EthTxHashManager) Revert(ctx context.Context, from, to *types.TipSet) e
 	return nil
 }
 
+// FillIndexGap back-fills the gap between the current head and the last populated transaction index
```
Comment (aarshkshah1992) [inline node/impl/full/txhashmanager.go:26]: nit: "last populated transaction index in* the Index"

File: node/impl/full/txhashmanager.go [hunk 5]
```diff
@@ -23,6 +23,49 @@ func (m *EthTxHashManager) Revert(ctx context.Context, from, to *types.TipSet) e
...
+	lastTxHash, lastTxCid, err := m.TransactionHashLookup.GetLastTransaction()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Start filling transaction index gap from: %s", lastTxHash)
+
+	totalProcessedBlock := 0
+	ts := m.StateAPI.Chain.GetHeaviestTipSet()
+	for _, block := range ts.Blocks() {
```
Comment (aarshkshah1992) [inline node/impl/full/txhashmanager.go:37]: I think we need to keep going back in the chain to backfill. This loop terminates after backfilling messages in the heaviest tipset only.

File: node/impl/full/txhashmanager.go [hunk 6]
```diff
@@ -23,6 +23,49 @@ func (m *EthTxHashManager) Revert(ctx context.Context, from, to *types.TipSet) e
...
+			}
+
+			// process the message
+			m.ProcessSignedMessage(ctx, msg)
+		}
+
+		totalProcessedBlock++
+	}
+
+	ts, err = m.StateAPI.Chain.GetTipSetFromKey(ctx, ts.Parents())
```
Comment (aarshkshah1992) [inline node/impl/full/txhashmanager.go:58]: This needs to move to the loop above

File: node/impl/full/txhashmanager.go [hunk 7]
```diff
@@ -23,6 +23,49 @@ func (m *EthTxHashManager) Revert(ctx context.Context, from, to *types.TipSet) e
...
+		if err != nil {
+			// If we can't find the messages, we've either imported from snapshot or pruned the store
+			log.Debug("exiting message mapping population at epoch ", ts.Height())
+			return nil
+		}
+
+		for _, msg := range msgs {
+			if msg.Cid() == lastTxCid {
+				// We've reached the last indexed transaction
+				return nil
```
Comment (aarshkshah1992) [inline node/impl/full/txhashmanager.go:48]: log completion and then return. Add logging around how many entires were inserted, which tipset it started from and which one it ended at etc.

**Review by aarshkshah1992 — COMMENTED** [review]
Round 1 of review. @akaladarshi Please can we also add logic to populate this from a snapshot just like we do for the `MsgIndex` ?

File: node/impl/full/txhashmanager.go (hunk 5 above)
Comment (rvagg) [reply node/impl/full/txhashmanager.go:37]: We need to be careful on two fronts with going backward up the chain for two reasons:

1. the last processed message CID may not be on the canonical chain anymore so you may never encounter it; and
2. the last processed message may not even be in the blockstore due to a splitstore compaciton or some other _happening_ (e.g. you've copied across someone else's txhash.db or restored an old one from disk, whatever)

I suspect `ChainGetPath` might be a good API to try here: get the tipset of the `lastTxHash`, get the latest tipset, find the path between. I'm just not sure what the error conditions are for that - how big a gap is too large, and how do we deal with splitstore GC problems (I suspect "find me the tipset for `lastTxHash` would error in that case if it was too old, in which case what's the strategy then?).

Comment (akaladarshi) [reply node/impl/full/txhashmanager.go:37]: @rvagg I have been going through the `ChainGetPath` code, but I didn't see it returning a `too old` error, it only returns a `not found` error.

If it is not found then nodes have to import the snapshot.
Is there any other way to get the state?

Also `ChainGetPath` returns both `revert` and `apply` changes, so do we have to filter out `apply` changes and start backfilling?

**Comment by akaladarshi** [comment]
@aarshkshah1992 For the observer part, I thought we could do something like this:

- Create a `backfill chan`
- Provide the `backfill chan` to the observer 
  - Once it gets  a new block from `ChainNotify` is sends it to the `backfill chan`
   - Wait till the `backfill chan` is closed (means backfill is done)
- Provide the same `backfill chan` to the `FillIndexGap` function in `EthTxManager`, also make it async
  - It will wait till it receives the new block through the `backfill chan` from the observer
  - Starts backfilling 
  - Once backfilling is done close the `backfill chan`  so the observer can proceed as usual

File: node/impl/full/txhashmanager.go (hunk 5 above)
Comment (aarshkshah1992) [reply node/impl/full/txhashmanager.go:37]: Discuss this with @akaladarshi  async. We worked through multiple solutions here. He will post a note about it.

File: node/config/types.go (hunk 2 above)
Comment (akaladarshi) [reply node/config/types.go]: Done.

**Pull request closed without merging**